File: app-flutter/lib/core/network/api_constants.dart
/// API Constants
/// All base URLs and endpoints centralized here
/// No hardcoded URLs anywhere else in the application
class ApiConstants {
  ApiConstants._();

  // ── Base URL — قابل للتجاوز عند الإصدار ─────────────────────────────
  //
  // **لماذا لا يكون ثابتاً صلباً؟**
  //   * خادم Render Free يدخل Sleep بعد 15 دقيقة من عدم الاستخدام، وأوّل
  //     طلب يستغرق 30–60 ثانية (cold start). نريد القدرة على نقل الخدمة
  //     لاحقاً إلى Render Paid أو Cloud Run بدون إصدار جديد ومراجعة متجر.
  //   * عند الترقية إلى Firebase Remote Config (مستقبليّاً) فقط أضف
  //     `firebase_remote_config` ومرّر النتيجة إلى `setOverride()` أثناء
  //     الـ bootstrap. كلّ الـ datasources تقرأ من `ApiConstants.baseUrl`
  //     بلا حاجة لتعديل.
  //
  // طريقة الاستخدام عند البناء:
  //   `flutter build apk --dart-define=NEBRAS_API_BASE_URL=https://api.example.com/`
  static const String _defaultBaseUrl = 'https://nebras-backend.onrender.com/';
  static const String _envBaseUrl = String.fromEnvironment(
    'NEBRAS_API_BASE_URL',
    defaultValue: '',
  );
  static String? _runtimeOverride;

  /// يستخدم في الـ bootstrap (مثلاً بعد قراءة Firebase Remote Config) لضبط
  /// عنوان الخادم في وقت التشغيل دون إعادة بناء. مرّر `null` للعودة إلى
  /// القيمة الافتراضيّة.
  static void setOverride(String? url) {
    _runtimeOverride = (url != null && url.trim().isNotEmpty)
        ? url.trim()
        : null;
  }

  /// يُعيد عنوان الـ API الحاليّ بحسب الأولويّة:
  ///   1. runtime override (مثلاً من Remote Config)
  ///   2. `--dart-define=NEBRAS_API_BASE_URL=...` عند البناء
  ///   3. الافتراضيّ المضمَّن في الكود
  static String get baseUrl {
    if (_runtimeOverride != null) return _runtimeOverride!;
    if (_envBaseUrl.isNotEmpty) return _envBaseUrl;
    return _defaultBaseUrl;
  }

  // Timeouts
  static const Duration connectTimeout = Duration(seconds: 30);
  static const Duration receiveTimeout = Duration(seconds: 60);
  static const Duration sendTimeout = Duration(seconds: 30);

  // Retry configuration
  static const int maxRetries = 3;
  static const Duration retryDelay = Duration(seconds: 2);
  //content
  static const String content = '/api/public/all/';
  static String contentDetail(int id) => '$content$id/';

  // ── Sections ────────────────────────────────────────────────
  static const String mainSections = '/api/public/sections/main/';
  static String mainSectionDetail(String name) => '$mainSections$name/';
  static const String subSections = '/api/public/sections/sub/';
  static String subSectionsByMain(String mainName) =>
      '$subSections?main_section=$mainName';
  static String subSectionDetail(String name) => '$subSections$name/';
  static const String secondarySections = '/api/public/sections/secondary/';
  static String secondarySectionsBySub(String subId) =>
      '$secondarySections?sub_section=$subId';
  // ── Content Metadata ────────────────────────────────────────
  static const String metadata = '/api/public/metadata/';

  // ── YouTube ─────────────────────────────────────────────────
  static const String youtube = '/api/public/youtube/';
  static String youtubeDetail(int id) => '$youtube$id/';

  // ── Files (R2) ──────────────────────────────────────────────
  static const String files = '/api/public/files/';
  static String fileDetail(int id) => '$files$id/';

  // ── Notifications ──────────────────────────────────────────
  static const String notifications = '/api/notifications/devices/';
  static String notificationDetail(int id) => '$notifications$id/';
}

File: nebras-app/app/src/main/kotlin/com/nebras/mobile/core/network/ApiConstants.kt
package com.nebras.mobile.core.network

/**
 * كلّ عناوين الـ API ونقاط النهاية مركزيّة هنا — لا رابط مكتوب يدويّاً في
 * أيّ مكان آخر من التطبيق.
 */
object ApiConstants {

    // ── Base URL — قابل للتجاوز عند التشغيل ─────────────────────────────
    //
    // **لماذا لا يكون ثابتاً صلباً؟**
    //   * خادم Render Free يدخل Sleep بعد 15 دقيقة، وأوّل طلب يستغرق
    //     30–60 ثانية (cold start). نريد نقل الخدمة لاحقاً إلى Cloud Run
    //     بدون إصدار جديد ومراجعة متجر.
    //   * عند الترقية إلى Firebase Remote Config مرّر النتيجة إلى
    //     [setOverride] أثناء الـ bootstrap؛ كلّ الـ datasources تقرأ من
    //     [baseUrl] بلا حاجة لتعديل.
    private const val DEFAULT_BASE_URL = "https://nebras-backend.onrender.com/"

    @Volatile
    private var runtimeOverride: String? = null

    /**
     * يُستخدم في الـ bootstrap (مثلاً بعد قراءة Remote Config) لضبط عنوان
     * الخادم وقت التشغيل دون إعادة بناء. مرّر `null` للعودة إلى الافتراضيّ.
     */
    fun setOverride(url: String?) {
        runtimeOverride = url?.trim()?.takeIf { it.isNotEmpty() }
    }

    /**
     * عنوان الـ API الحاليّ بحسب الأولويّة:
     *   1. runtime override (مثلاً من Remote Config)
     *   2. الافتراضيّ المضمَّن في الكود
     */
    val baseUrl: String
        get() = runtimeOverride ?: DEFAULT_BASE_URL

    // ── المهلات ─────────────────────────────────────────────────────────
    // للعميل العامّ (طلبات JSON فقط). مسارات التنزيل/الوسائط الكبيرة
    // (audio/video/reader/download) تضبط مهلاتها الخاصّة (دقائق) ولا تعتمد
    // على هذه القيم.
    //
    // connectTimeout = 12s: على إنترنت ضعيف يفشل ربط TCP/TLS بسرعة فيبدأ
    // [RetryInterceptor] المحاولة بدل تعليق المستخدم 30 ثانية. إقلاع الخادم
    // البطيء يُغطّى بإعادة المحاولة لا بمهلة ربط طويلة.
    const val CONNECT_TIMEOUT_SECONDS = 12L
    const val RECEIVE_TIMEOUT_SECONDS = 30L
    const val SEND_TIMEOUT_SECONDS = 30L

    // ── إعادة المحاولة ──────────────────────────────────────────────────
    // RETRY_DELAY يُستعمل كـ«قاعدة» للتراجع الأُسّي مع jitter
    // (≈2s ثمّ 4s ثمّ 8s) بدل تأخير ثابت.
    const val MAX_RETRIES = 3
    const val RETRY_DELAY_MILLIS = 2_000L

    // ── المحتوى ─────────────────────────────────────────────────────────
    const val CONTENT = "/api/public/all/"
    fun contentDetail(id: Int): String = "$CONTENT$id/"

    // ── الأقسام ─────────────────────────────────────────────────────────
    const val MAIN_SECTIONS = "/api/public/sections/main/"
    fun mainSectionDetail(name: String): String = "$MAIN_SECTIONS$name/"

    const val SUB_SECTIONS = "/api/public/sections/sub/"
    fun subSectionsByMain(mainName: String): String = "$SUB_SECTIONS?main_section=$mainName"
    fun subSectionDetail(name: String): String = "$SUB_SECTIONS$name/"

    const val SECONDARY_SECTIONS = "/api/public/sections/secondary/"
    fun secondarySectionsBySub(subId: String): String = "$SECONDARY_SECTIONS?sub_section=$subId"

    // ── بيانات وصفيّة ───────────────────────────────────────────────────
    const val METADATA = "/api/public/metadata/"

    // ⛔ نقاط YouTube بقيت في العقد الخادميّ لكنّ التطبيق لا يستدعيها إطلاقاً
    // (YouTube مقطوع نهائياً — انظر ContentCompliance).

    // ── الملفّات (R2) ───────────────────────────────────────────────────
    const val FILES = "/api/public/files/"
    fun fileDetail(id: Int): String = "$FILES$id/"

    // ── الإشعارات ───────────────────────────────────────────────────────
    const val NOTIFICATIONS = "/api/notifications/devices/"
    fun notificationDetail(id: Int): String = "$NOTIFICATIONS$id/"
}
